bot/Engine/BotController.py

```diff
@@ -0,0 +1,385 @@
+            # How can we guarantee timestamp of exchange is always in ms?
+            if order.timestamp - time.time()*1000 > self.bot.entry_settings.buy_order_time_out:
+                if self.test_mode:
+                    order_result = exchange.cancelOrder(order.symbol, order.id)
+                else:
+                    order_result = dict(status=self.exchange.ORDER_STATUS_CANCELED)
+                if exchange.isValidResponse(order_result):
+                    order.status = order_result['status']
+                    self.processClosedPosition(order, pair)
+
+    def updateOpenSellOrder(self, order, pair):
+        """
+        Update open sell order based on given params in exit settings. 
+        For now only supports placing stop loss market order when position goes against us ( basically simple oco order variant ).
+        """
+        exchange = self.exchange
+        bot = self.bot
+        strategy = self.strategy
+        
+        candlestick_data = exchange.getSymbolKlines(order.symbol, self.kline_interval, strategy.minimum_period)
+        current_price = candlestick_data.iloc[-1]['close']
+        stop_loss_active = bot.exit_settings.stop_loss_value is not None
+
+        if stop_loss_active:
+            if ((Decimal(100) - Decimal(bot.exit_settings.stop_loss_value))/Decimal(100)) * Decimal(order.entry_price) >= Decimal(current_price):
+                quantity = self.computeQuantity(order)
+                side = 'SELL'
+                order_type = exchange.ORDER_TYPE_MARKET
+
+                if not self.test_mode:
+                    order_result = exchange.cancelOrder(order.symbol, order.id)
+                else:
+                    order_result = dict(status=self.exchange.ORDER_STATUS_CANCELED, executedQty=0)
+
+                if exchange.isValidResponse(order_result):
+                    # TODO probably taking over info from exchange like this will not be modular enough for other exchanges.
+                    order.status = order_result['status']
+                    order.executed_quantity = order_result['executedQty']
+
+                    self.placeNewOrder( 
+                                        symbol=order.symbol, 
+                                        pair=pair,
+                                        order=order,
+                                        quantity=quantity,
+                                        order_type=order_type,
+                                        side=side
+                                        ) 
+	
+    def tryExitOrder(self, order, pair):
+        """ If strategy returns exit signal look to place exit order. """
+        bot = self.bot
+        symbol = order.symbol
+        exchange = self.exchange
+        strategy = self.strategy
+
+        candlestick_data = exchange.getSymbolKlines(symbol, self.kline_interval, strategy.minimum_period)
+        exit_signal = strategy.shouldExitOrder(candlestick_data)
+
+        if exit_signal:
```

Comment:
Here, it depends whether our exit settings say to exit on signal, or exit on take profit. Will need to change it a little bit, but leave it to me... need to do some thinking first. The step that checks the exit settings should be before exit_signal is checked - and exit_signal should only be checked if exit_settings warrants that.

Reply:
I think I do get your line of thinking here, if we want to exit on a signal we just want a market_order (or maybe some limit_order at some distance from price at that point). Stop-loss and take-profit are kind of useless if we have an exit on signal. We should create an extra column in the models for exit on signal.

Atm the checkToExitPosition() is returns True in strategies if no exit on signal is specified, therefore, my line of thinking was that conditioned on what is specified there this would work. But I get why this does not make sense.